cmake/Dependencies.cmake

```diff
@@ -274,44 +274,59 @@ if(INTERN_BUILD_MOBILE AND INTERN_USE_EIGEN_BLAS)
 endif()
 
 # ---[ pthreadpool
-# QNNPACK and NNPACK both depend on pthreadpool, but when building with libtorch
-# they should use the pthreadpool implementation under caffe2/utils/threadpool
-# instead of the default implementation. To avoid confusion, add pthreadpool
-# subdirectory explicitly with EXCLUDE_FROM_ALL property prior to QNNPACK/NNPACK
-# does so, which will prevent it from installing the default pthreadpool library.
-if(INTERN_BUILD_MOBILE AND NOT BUILD_CAFFE2_MOBILE AND (USE_QNNPACK OR USE_NNPACK OR USE_XNNPACK))
-  if(NOT DEFINED PTHREADPOOL_SOURCE_DIR)
-    set(CAFFE2_THIRD_PARTY_ROOT "${PROJECT_SOURCE_DIR}/third_party")
-    set(PTHREADPOOL_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/pthreadpool" CACHE STRING "pthreadpool source directory")
-  endif()
+# pthreadpool is a C library with no concept of namespaces, which means two
+# copies of the library cannot exist in the same binary or symbol collision
+# will occur violating ODR.  The conundrum is that Caffe2 provides its own
+# custom implementation of pthreadpool under the exact same interface the open
+# source version operates under; an implementation that is only used in Caffe2
+# operators and Caffe2's usage of NNPACK and QNNPACK. On the other hand, we
+# are interested in using the open source implementation on PyTorch which seems
+# to have been updated to exhibit superior performance.  Here we try to solve
+# this conundrum by prefering to pick the internal Caffe2 implementation if
+# BUILD_CAFFE2_OPS is on.  This will force both PyTorch and Caffe2 to use the
+# Caffe2 pthreadpool version (remember we cannot have two copies in one binary)
+# and in doing so we achieve our goal of keeping Caffe2's behavior intact if
+# at the cost of some performance loss to PyTorch.  Otherwise, if BUILD_CAFFE2_OPS
+# is not requested, we take the liberty of picking the open source version to
+# improve PyTorch's performance.  All of this admittedly error prone logic
+# can be removed and simplified if and when we decide to migrate Caffe2 to the
+# open source pthreadpool as well.
+
+if(NOT USE_SYSTEM_PTHREADPOOL)
+  # Opt for custom Caffe2 implementation whenever BUILD_CAFFE2_OPS is enabled
+  if(BUILD_CAFFE2_OPS)
```

Comment:
I think this is really gonna matter for internal build system? I am not sure if it matters to keep this here. It is little strange now that you cannot use xnnpack if you turn this flag ON.
For internal build I suppose we can set this for very specific builds for C2. But if we are building pytorch lib then we should not use this?
Also I am wondering if this is further going to complicate other things when we dispatch to C2 ops from PyTorch as I think Linbin was trying. @linbinyu.
My apologies here. I know this has already been complicated enough so dont want to throw more troubles here, but let's make sure to bring all the stakeholders here and make sure it does not build existing things.

Reply:
Thank you for your comment Kimish.  These are very valid points.  I think this is the trickiest part of this PR.  Since we cannot have two copies of pthreadpool, we have to somehow surgically decide when to use what.  Alternatively we can just use third-party pthreadpool in CMake for all OSS use cases, regardless of whether we use C2 or PT, and only concentrate on making the C2 vs PT separation for internal usecases.  As a matter of fact that would make things mich easier.  The reason a number of builds are currently failing is because I need to modify the JNI builds to only link against third_party pthreadpools and XNNPACK, only if they are available which is no longer the case with these changes.

@dreiss What do you think?

Reply:
I think it's fine to move everything to the new third party pthreadpool for all OSS builds.

It looks like none of the ops that Linbin will be affected.  GenerateProposals doesn't appear to use threads, and RoIAlign only uses OpenMP.